cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java

```diff
@@ -367,8 +367,41 @@ private Builder(Schema schema) {
      * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
      *                                   value is given
      */
+
+    private boolean checkAndSetLogicalSchema(Schema schema, String fieldName, @Nullable Object value) {
+      schema = getLogicalTypeSchema(schema, EnumSet.of(LogicalType.FIXED));
+      if (schema == null) {
+        return false;
+      }
+      if (schema.getLogicalType().equals(LogicalType.FIXED)) {
+        setFixedLogicalTypeSchema(schema, fieldName, value);
+        return true;
+      }
+      return false;
+    }
+    private void setFixedLogicalTypeSchema(Schema schema, String fieldName, @Nullable Object value) {
+      if (schema.getLogicalType().equals(LogicalType.FIXED)) {
+        ByteBuffer byteBuffer = null;
+        if (value instanceof ByteBuffer) {
+          byteBuffer = (ByteBuffer) value;
+        } else if (value instanceof byte[]) {
+          byteBuffer = ByteBuffer.wrap((byte[]) value);
+        }
+        if (byteBuffer.remaining() !=  schema.getFixedSize()) {
+          throw new UnexpectedFormatException(
+            String.format("Field '%s' has size '%d' which is not equal to schema size '%d'.",
+              fieldName, byteBuffer.remaining(), schema.getFixedSize()));
+        }
+        fields.put(fieldName, byteBuffer);
+      }
+
+    }
     public Builder set(String fieldName, @Nullable Object value) {
-      validateAndGetField(fieldName, value);
+      Schema.Field field = validateAndGetField(fieldName, value);
+      if (value != null && field != null && field.getSchema().getLogicalType() != null &&
```

Comment:
nit: put the `&&` at the start of the second line, as it makes evident that this is a continuation of the above condition.

Reply:
```
if (value != null && field != null && field.getSchema().getLogicalType() != null 
    && checkAndSetLogicalSchema(field.getSchema(), fieldName, value)) {
    return this;
}
```

Reply:
Done